### PR TITLE
Add semantic search with Voyage AI embeddings

### PR DESCRIPTION
**Pull request by user1303836**
## Summary

- Add `/search` slash command for semantic search across all stored content (Substack, YouTube, RSS, ArXiv, Blog, Twitter, Page)
- Use Voyage AI `voyage-3.5-lite` embeddings with in-memory numpy cosine similarity for ranking
- Embeddings generated incrementally during pipeline cycles via `embed_pending()`, with automatic backfill of existing content
- New `ContentEmbedding` database model with ORM cascade deletion chain (Source -> ContentItem -> ContentEmbedding)
- `EmbeddingProvider` protocol for swappable embedding backends
- Feature activates when `VOYAGE_API_KEY` is configured; gracefully disabled otherwise
- Configurable similarity threshold, max results, and embedding model via settings
- Automatic re-embedding when embedding model changes
- Guild-scoped search results with optional source type and date range filters
- 60 new tests covering service, cog, and database layers (633 total, all passing)

## Key files

| File | Purpose |
|------|---------|
| `src/intelstream/services/search.py` | SearchService, EmbeddingProvider protocol, VoyageEmbeddingProvider |
| `src/intelstream/discord/cogs/search.py` | SearchCog with /search command |
| `src/intelstream/database/models.py` | ContentEmbedding model |
| `src/intelstream/database/repository.py` | Embedding CRUD and search query methods |
| `src/intelstream/services/pipeline.py` | Pipeline integration (embed_pending in run_cycle) |
| `src/intelstream/config.py` | voyage_api_key, search settings |

## Test plan

- [x] All 633 tests pass (573 existing + 60 new)
- [x] ruff check passes
- [x] ruff format passes
- [x] mypy strict mode passes
- [x] Existing content_posting tests still pass with mock voyage_api_key=None
- [x] Database cascade deletion verified (source delete removes embeddings)
- [x] Model consistency check verified (model change clears all embeddings)

### CODE REVIEW

File: src/intelstream/discord/cogs/search.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            await interaction.followup.send(
+                "An error occurred while searching. Please try again.", ephemeral=True
+            )
+            return
+
+        if not results:
+            await interaction.followup.send("No relevant results found.", ephemeral=True)
+            return
+
+        embed = discord.Embed(
```
Comment (user1303836) [inline src/intelstream/discord/cogs/search.py]: **[Blocking]** The query string is embedded directly into the Discord embed title without sanitization. While Discord embeds don't execute code, a very long query (up to 200 chars) combined with the `Search: "..."` prefix could exceed the 256-char embed title limit. The title format is `f'Search: "{query}"'` which adds 11 characters of overhead, so queries up to 200 chars would produce titles up to 211 chars -- under the 256 limit.

Actually, re-checking: 200 + 11 = 211 < 256, so this is fine. Disregard this concern.

However, the `result.title[:250]` truncation on line 93 could produce field names up to 253 chars (`"1. " + 250`), which also fits under the 256-char embed field name limit. Good.

One real concern: if `result.original_url` is very long, the field value on line 94 could exceed the 1024-char embed field value limit. Consider truncating the URL or the overall field value.

File: src/intelstream/database/repository.py [hunk 2]
```diff
@@ -890,3 +891,80 @@ async def set_github_repo_active(self, repo_id: str, is_active: bool) -> bool:
...
+    async def get_embeddings_with_items(
+        self,
+        guild_id: str | None = None,
+        source_type: str | None = None,
+        since: datetime | None = None,
+    ) -> list[tuple[ContentEmbedding, ContentItem, Source]]:
+        async with self.session() as session:
+            query = (
+                select(ContentEmbedding, ContentItem, Source)
+                .join(ContentItem, ContentEmbedding.content_item_id == ContentItem.id)
```
Comment (user1303836) [inline src/intelstream/database/repository.py]: **[Blocking]** `SourceType(source_type)` here will raise `ValueError` if the user passes an invalid source type string that doesn't map to a `SourceType` enum value. This would cause an unhandled exception that bubbles up through the search path.

The search cog uses `@app_commands.choices` which constrains the Discord UI, but a malformed API request could still pass an invalid string. The exception would be caught by the cog's broad `except Exception` handler, which would show a generic error message -- acceptable but not ideal for UX.

Consider wrapping in a try/except in either the repository method or the `_get_candidates` caller:
```python
try:
    query = query.where(Source.type == SourceType(source_type))
except ValueError:
    return []  # invalid source_type = no results
```

File: src/intelstream/services/search.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                    text_hash=text_hash,
+                )
+                self._embeddings_cache[item.id] = np.array(embeddings[0], dtype=np.float32)
+                count += 1
+            except Exception:
+                logger.warning("Failed to embed content item", item_id=item.id, exc_info=True)
+                continue
+        return count
+
+    async def _ensure_cache_loaded(self) -> None:
```
Comment (user1303836) [inline src/intelstream/services/search.py]: **[Blocking]** `_ensure_cache_loaded()` loads ALL embeddings from the database into memory on first search. This is correct per the design, but there's a subtle race condition: if `embed_pending()` is running concurrently (from the pipeline cycle) while a user triggers a search, the cache could be loaded in an incomplete state. 

Since Python's asyncio is single-threaded, this isn't a real concurrency issue -- the `await` points won't interleave within a single coroutine's logic. So this is actually fine. Disregard.

However, there IS a real issue: after `_check_model_consistency()` clears the cache and sets `_cache_loaded = False` (line 154), the subsequent `_get_candidates()` call will reload the cache. But if model consistency was just checked and all embeddings were cleared, the reload will load an empty set. Then `embed_pending()` will re-embed items, adding them to the cache. On the NEXT search, `_cache_loaded` is already `True`, so the cache won't reload -- but the cache was populated incrementally by `embed_pending()`, so this actually works correctly.

OK, I've convinced myself the logic is sound. No change needed.

**Review by user1303836 — COMMENTED** [review]
Thorough review of the semantic search implementation. Overall this is solid work -- clean architecture, good test coverage, and proper integration with existing patterns.

However, there are a few blocking issues that need to be addressed before merge:

**Blocking Issues:**

1. **Zero-norm vector bug in `cosine_similarity`** (`search.py:65`) - Division by zero when a zero-norm vector is passed. This can happen if an embedding API returns all zeros or if data gets corrupted. Add a guard.

2. **`_check_model_consistency()` called on every search** (`search.py:102`) - This hits the DB on every single search query. Should be called once during cache load or on a timer, not on the hot path.

3. **No validation on `days` parameter** (`cogs/search.py:45`) - Zero or negative values would produce nonsensical results. Add a minimum bound check.

4. **Unhandled `ValueError` from `SourceType(source_type)`** (`repository.py:945`) - If an invalid source_type string is passed, this raises an unhandled ValueError. Needs a try/except or pre-validation.

5. **Missing zero-vector test** (`test_search.py`) - Add a test for the zero-vector edge case in `cosine_similarity` to prevent regression.

**Suggestions (non-blocking):**
- Consider batching embedding API calls in `embed_pending` instead of one call per item
- Add a `setup()` function at the bottom of `cogs/search.py` for pattern consistency with other cogs
- The hardcoded `dimension = 1024` in VoyageEmbeddingProvider could be a lookup table keyed by model name

See inline comments for details.
